### DISCUSSION TITLE
How to programmatically subscribe to custom Debug Adapter events?

### DISCUSSION THREAD

**Discussion by sebthom** [eclipse/lsp4e#652] [Q&A] [answered]
I am launching a debug adapter via DSPLaunchDelegateLaunchBuilder, see https://github.com/dart4e/dart4e/blob/6a925b7078089f9810cbedb22b819789e3e8bc63/plugin/src/main/java/org/dart4e/launch/program/ProgramLaunchConfigLauncher.java#L90-L98

Now I am trying to figure out how I can get notified of these custom debug adapter events https://github.com/dart-lang/sdk/blob/main/pkg/dds/tool/dap/README.md#custom-events

Any help is greatly appreciated!

@jonahgraham

**Comment by jonahgraham** [comment, marked as answer]
Hi @sebthom - not sure what you have tried yet here, but my guess is you want to override [`createDebugTarget`](https://github.com/eclipse/lsp4e/blob/73d8dc2f6541ad5303f7aab945e01e7f746ab934/org.eclipse.lsp4e.debug/src/org/eclipse/lsp4e/debug/launcher/DSPLaunchDelegate.java#L386) in [`LaunchDebugConfig`](https://github.com/dart4e/dart4e/blob/6a925b7078089f9810cbedb22b819789e3e8bc63/plugin/src/main/java/org/dart4e/launch/LaunchDebugConfig.java#L33). Have `createDebugTarget` return a specialization of [`DSPDebugTarget`](https://github.com/eclipse/lsp4e/blob/73d8dc2f6541ad5303f7aab945e01e7f746ab934/org.eclipse.lsp4e.debug/src/org/eclipse/lsp4e/debug/debugmodel/DSPDebugTarget.java) that implements a new interface extending [`IDebugProtocolClient`](https://github.com/eclipse-lsp4j/lsp4j/blob/e32c63b7b1cea5b777bd0d99c4045f4932ad3ff7/org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/services/IDebugProtocolClient.java).

That new interface can have methods sort of like this which get implemented in your new class:

```java
	@JsonNotification("dart.log")
	default void dartLog(String message) {
	}
```

Please ask follow up questions if the above isn't clear to what you expected.

**Reply by sebthom** [reply]
Perfect, thanks!
